torch/csrc/jit/codegen/cuda/predicate_compute.cpp

```diff
@@ -106,6 +107,82 @@ std::vector<kir::Bool*> PredicateCompute::computePredicates(
   return preds;
 }
 
+namespace {
+
+//! Analyze whether IterDomain can be statically determined to be safe
+//! without bounds-checking predicates.
+class IterationDomainAnalysis : private ExpressionEvaluator {
+ public:
+  //! Return true if the expression defining tv can be safely run
+  //! without a predicate
+  static bool canOmitPredicate(const kir::TensorView* tv) {
+    const auto gpu_lower = GpuLower::current();
+    auto fuser_tv = tv->fuserTv();
+    for (size_t i = 0; i < fuser_tv->nDims(); ++i) {
+      IterDomain* id =
+          gpu_lower->caLoopMap().getConcreteMappedID(fuser_tv->axis(i));
+      IterationDomainAnalysis id_analysis(id->fusion());
+      auto extent = id->rawExtent();
+      id_analysis.evaluate(extent);
+      if (!id_analysis.isExact(extent)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+ private:
+  IterationDomainAnalysis(Fusion* fusion) : ExpressionEvaluator(fusion) {}
+
+  using ExpressionEvaluator::handle;
+
+  //! Check if val has nothing that prevents a loop using val as its
+  //! extent to omit a bounds-checking predicate
+  bool isExact(const Val* val) {
+    return val->definition() == nullptr ||
+        exact_vals_.find(val) != exact_vals_.end();
+  }
+
+  //! Record val does not need a predicate.
+  void setExact(const Val* val) {
+    exact_vals_.insert(val);
+  }
+
+  void handle(BinaryOp* bop) override {
+    ExpressionEvaluator::handle(bop);
+
+    const auto lhs = bop->lhs();
+    const auto rhs = bop->rhs();
+
+    if (!(isExact(lhs) && isExact(rhs))) {
+      return;
+    }
+
+    if (bop->getBinaryOpType() == BinaryOpType::CeilDiv) {
+      // CeilDiv is the only expression that can make an extent val
+      // larger than the actual. Need to know the exact values.
+      const auto lhs_value = getValue(lhs);
+      const auto rhs_value = getValue(rhs);
+      if (lhs_value.has_value() && rhs_value.has_value() &&
+          (lhs_value.value() % rhs_value.value()) == 0) {
+        setExact(bop->out());
+      }
+    } else if (bop->getBinaryOpType() == BinaryOpType::Mul) {
+      setExact(bop->out());
+    } else {
+      // Expr on extent should be either CeilDiv or Mul, which are
+      // derived from split and merge, respectively.
+      TORCH_INTERNAL_ASSERT("Unexpected BinaryOpType: ", bop);
+    }
+  }
+
+ private:
+  //! Vals that are known to need no predicate if used as IterDomain extent
+  std::unordered_set<const Val*> exact_vals_;
```

Comment:
we can fold this into the ExpressionEvaluator itself

Reply:
This is very specific to this particular logic here, so I don't think folding it into ExpressionEvaluator is a good idea.

Reply:
I'll admit I don't fully follow the logic, which indeed seems very specialized (what happens if we have anything other than  mul or ceildiv in the expression - even a harmless x+0 appears to throw off the current implementation.

Maybe we're not too far off: my point is that either we deal with a general expression evaluation problem, case in which it makes sense to abstract the logic and reuse EE, or it's not a general EE problem then we should not contort EE just to save a few lines of code.